### PR TITLE
Updates for 2023

### PR DESCRIPTION
**Pull request by rcunrau**
- antlr4 updated to 4.13.0
- cmake configures and builds MLIR
- added BackEnd class to generate a main function that calls `printf("\n")`

### CODE REVIEW

File: README.md
Comment (quinnlp) [inline README.md:1]: We'll also need to change the install instructions here since we want them to install a newer version of LLVM from the project directly (no fork), and build MLIR. I can write this up later.

Comment (kuzi117) [reply README.md:1]: Gotta be careful with his. The reason we forked is because the main repo updates their branches with backported fixes. So if you told people to check out the 10.0 branch from the main repo, there's a case where people are actually at different hashes. It's _unlikely_ that any of their backported fixes significantly affect anything the students are doing/working with and I seriously doubt there'd every be an interface change, but the possibility exists of different code gen/results.

Basically I just want to make sure you're either aware of this or telling them to check out by hash and not tag/branch.

Comment (rcunrau) [reply README.md:1]: Interesting. I would have thought that with a new release every month (or thereabouts) they'd be forward looking, i.e. if there's a bug in this release just update to the next release. Even more surprising that they would back port fixes, since again I never saw any mention of support or LTS. Do they have any statement of their support policies anywhere? Anyway thanks for the warning, we'll discuss it and try to come up with a plan

Comment (kuzi117) [reply README.md:1]: For all I know they've changed this policy, I haven't been following them in a while. I know they've changed release policies at least once before in a large way. I'll check and see if I can see any evidence of them still doing this.

Comment (rcunrau) [reply README.md:1]: The aggressive release plan is MLIR, of course. LLVM is used in tons of compilers so of course they have support, but it's also more stable.

Comment (kuzi117) [reply README.md:1]: So it seems like the branches are still changing and they're different from the main line. So the [16.x](https://github.com/llvm/llvm-project/commits/release/16.x) branch has specific fixes as well as a lot of things cherry picked from main. However, it also seems like the tags, which can't change, are plucked off these release branches. You should be able to select something like [llvmorg-16.0.6](https://github.com/llvm/llvm-project/commits/llvmorg-16.0.6) and have that be consistent.

Comment (rcunrau) [reply README.md:1]: WRT setup procedures in VCalc and Gazprea, I looks like cmake only needs MLIR_DIR set to be able to create a working set of Makefiles. I think this is correct, specifically that VCalc and Gazprea don't need to know where MLIR comes from or what version it is. Setting up MLIR should be in the Setup instructions so we keep all the version chaos in one place. Actually, it would be nice to setup the ANTLR instructions this way too, so we don't have to change 8 files every time we update it.

Comment (rcunrau) [reply README.md:1]: I looked at the 16.x branch that you mentioned, and yes, they are updating it with fixes. However, when they release the new branch they give it a new number (e.g. 16.0.6) and that should have a new tag. We are not recommending the students check out a MLIR branch, we are recommending they check out a release using a tag. That tag should be the same as a hash and it should be the same for everyone. If we find a bug in that snapshot, then we'll give the whole class a new release tag to checkout. That way everyone should always have the same thing.

Comment (quinnlp) [reply README.md:1]: Thanks for the warning Braedy. We weren't sure why the the class had a forked version of llvm. And yes, our plan is to do this:
> You should be able to select something like [llvmorg-16.0.6](https://github.com/llvm/llvm-project/commits/llvmorg-16.0.6) and have that be consistent.

Comment (kuzi117) [reply README.md:1]: > WRT setup procedures in VCalc and Gazprea, I looks like cmake only needs MLIR_DIR set to be able to create a working set of Makefiles. 
I believe this is true. I was sort of hazy because it's been awhile, but I just double checked. It's an artifact of (cmake's find_package command)[https://cmake.org/cmake/help/v3.0/command/find_package.html] which looks for <package>_DIR in the session's environment variables as a place to find <package>Config.cmake. So as long as it's true that you're looking for MLIR in MLIR_DIR and MLIRConfig.cmake exists, everything should work. LLVM's been pretty consistent about their setup, so I expect they translated it directly to MLIR.

File: README.md [hunk 1]
```diff
@@ -17,43 +17,40 @@ in the scripts folder and understand as much as possible. We'll touch some
...
+     1. `git clone https://github.com/llvm/llvm-project.git`
+     1. `cd llvm-project`
+     1. `git checkout llvmorg-16.0.4`
+  1. Build MLIR
+     1. `mkdir build`
+     1. `cd build`
+     1. ```cmake -G Ninja ../llvm \
+             -DLLVM_ENABLE_PROJECTS=mlir \
+             -DLLVM_BUILD_EXAMPLES=ON \
+             -DLLVM_TARGETS_TO_BUILD="Native;NVPTX;AMDGPU" \
```
Comment (kuzi117) [inline README.md:31]: You're compiling for GPU? Also interesting that `Native` is option, I had no idea.

File: README.md [hunk 2]
```diff
@@ -17,43 +17,40 @@ in the scripts folder and understand as much as possible. We'll touch some
...
-     `~/.bash_profile` on MacOS to setup the next steps. You should restart your
+     1. `git clone https://github.com/llvm/llvm-project.git`
+     1. `cd llvm-project`
+     1. `git checkout llvmorg-16.0.4`
+  1. Build MLIR
+     1. `mkdir build`
+     1. `cd build`
+     1. ```cmake -G Ninja ../llvm \
+             -DLLVM_ENABLE_PROJECTS=mlir \
+             -DLLVM_BUILD_EXAMPLES=ON \
```
Comment (kuzi117) [inline README.md:30]: Are you sure you really want to build examples? I've found it to be a waste of build time in a lot of cases. Does it do something interesting for MLIR?

File: README.md [hunk 3]
```diff
@@ -17,43 +17,40 @@ in the scripts folder and understand as much as possible. We'll touch some
...
-      export LLVM_INS="$HOME/llvm-ins/" # Change me if you really want.
-      export LLVM_DIR="$LLVM_INS/lib/cmake/llvm/" # Don't change me.
-      export PATH="$LLVM_INS/bin:$PATH" # Don't change me
+      export MLIR_INS="$HOME/llvm-project/build/"
+      export MLIR_DIR="$MLIR_INS/lib/cmake/mlir/" # Don't change me.
+      export PATH="$MLIR_INS/bin:$PATH" # Don't change me
       ```
-  1. If you want a debug build you may change this by looking at Step 2 in the
-     `scripts/configureLLVM.sh` script.
-  1. Run `configureLLVM.sh`.
```
Comment (kuzi117) [inline README.md:37]: Is this script just fully cut? It's still in the repo though.

File: README.md (hunk 1 above)
Comment (rcunrau) [reply README.md:31]: This is from the MLIR getting started page (which I should probably reference in the readme). I tested it on my Mac (which doesn't have an AMDGPU).

File: README.md (hunk 2 above)
Comment (rcunrau) [reply README.md:30]: The important example for MLIR is called "Toy". Unfortunately, they build a new dialect, which is way beyond what we want to do, but they do have lots of good MLIR generation examples. The big thing about dialects is that you specify them in template files that are then parsed and used to generate classes and methods. You only get this generated code if you build the example, and there's lot of good stuff in the generated files. This is also why "git grep" isn't as good as "grep -R".

File: README.md (hunk 3 above)
Comment (rcunrau) [reply README.md:37]: I left it in because we may want to adapt it to MLIR. It's good because it explains all the different options that the simple one in README.md uses.

File: README.md (hunk 1 above)
Comment (kuzi117) [reply README.md:31]: > I tested it on my Mac (which doesn't have an AMDGPU).

This is how you get backends for cross compilling. Running with this would let your install produce code for your native architecture, Nvidia's PTX intermediate, and AMD GPUs (I don't know what this is), but wouldn't let you generate code for another backend such as PPC.

> A semicolon delimited list controlling which targets will be built and linked into llvm. The default list is defined as LLVM_ALL_TARGETS, and can be set to include out-of-tree targets. The default value includes: AArch64, AMDGPU, ARM, AVR, BPF, Hexagon, Lanai,
Mips, MSP430, NVPTX, PowerPC, RISCV, Sparc,
SystemZ, WebAssembly, X86, XCore. Setting this to "host" will only compile the host architecture (e.g. equivalent to specifying X86 on an x86 host machine) can significantly speed up compile and test times.

Apparently native is a synonym for host, so that should would, but I would definitely drop nvptx and amdgpu, I don't think anyone is targeting nvidia or amd gpus.

Comment (rcunrau) [reply README.md]: Updated the build targets to "Native" only
